pkg/syncproto/sync_proto.go

```diff
@@ -12,6 +12,141 @@
+// KV message, each containing one or more KV pairs.
+//
+// Once a complete snapshot has been sent, Typha sends a SyncStatus message with
+// its current sync status.  This is typically "InSync" but it may be another status,
+// such as "Resync" if Typha itself is resyncing with the datastore.
+//
+// At any point after the  handshake, Typha may send a Ping message, which Felix
+// should respond to as quickly as possible with a Pong (if Typha doesn't receive
+// a timely response it may terminate the connection).
+//
+// After the initial snapshot is sent, Typha sends KVs and SyncStatus messages
+// as new updates are received from the datastore.
+//
+//	+-------+                +-------+
+//	| Felix |                | Typha |
+//	+-------+                +-------+
+//	|                        |
+//	| connect                |
+//	|----------------------->|
+//	|                        | -------------------------------\
+//	|                        |-| accept, wait for ClientHello |
+//	|                        | |------------------------------|
+//	|                        |
+//	| ClientHello            |
+//	|----------------------->|
+//	|                        |
+//	|            ServerHello |
+//	|<-----------------------|
+//	|                        | ------------------------------------\
+//	|                        |-| start KV send & pinger goroutines |
+//	|                        | |-----------------------------------|
+//	|                        |
+//	|                KVs * n |
+//	|<-----------------------|
+//	|                        |
+//	|                   Ping |
+//	|<-----------------------|
+//	|                        |
+//	| Pong                   |
+//	|----------------------->|
+//	|                        |
+//	|               KVs  * n |
+//	|<-----------------------|
+//	|                        |
+//	|     SyncStatus(InSync) |
+//	|<-----------------------|
+//	|                        |
+//	|               KVs  * n |
+//	|<-----------------------|
+//	|                        |
+//
+// Wire format
+//
+// The protocol uses gob to encode messages.  Each message is wrapped in an Envelope
+// struct to simplify decoding.
+//
+// Key/value pairs are encoded as SerializedUpdate objects.  These contain the KV pair
+// along with the Syncer metadata about the update (such as its revision and update type).
+// The key and value are encoded to the libcalico-go "default" encoding, as used when
+// storing data in, for example, etcd.  I.e. the gob struct contains string and []byte
+// fields to hold the key and value, respectively.  Doing this has a some advantages:
+//
+//     (1) It avoids any subtle incompatibility between our datamodel and gob.
+//
+//     (2) It removes the need to register all our datatypes with the gob en/decoder.
+//
+//     (3) It re-uses known-good serialization code with known semantics around
+//         data-model upgrade.
```

Comment:
For the "known semantics around data-model upgrade", do you mean as in the "Upgrading the datamodel" section below?  If so please add "(Please see next section.)" here.  (Otherwise I think it sounds like you might be referencing some other knowledge that might (or not) be "well-known".)